nixos/modules/services/monitoring/watchdogd.nix

```
@@ -0,0 +1,113 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.watchdogd;
+
+  mkPluginOpts = plugin: {
+    enable = mkEnableOption "watchdogd plugin ${plugin}";
+    interval = mkOption {
+      type = types.ints.unsigned;
+      default = 300;
+      description = ''
+        Amount of seconds between every poll.
+      '';
+    };
+    logmark = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to log current stats every poll interval.
+      '';
+    };
+    warning = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The critical watermark level. Alert sent to log, followed by reboot or script action.
+      '';
+    };
+  };
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
+    package = mkPackageOptionMD pkgs "watchdogd" { };
+
+    timeout = mkOption {
+      type = types.ints.unsigned;
+      default = 15;
+      description = ''
+        The WDT timeout before reset.
+      '';
+    };
+    interval = mkOption {
+      type = types.ints.unsigned;
+      default = 5;
+      description = ''
+        The kick interval, i.e. how often {manpage}`watchdogd(8)` should reset the WDT timer.
+      '';
+    };
+
+    safeExit = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        With {var}`safeExit` enabled, the daemon will ask the driver to disable the WDT before exiting.
+        However, some WDT drivers (or hardware) may not support this.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Additional config to put in {file}`watchdogd.conf`.
+        See {manpage}`watchdogd.conf(5)` for syntax.
+     '';
+    };
+
+    filenr = mkPluginOpts "filenr";
+    loadavg = mkPluginOpts "loadavg";
+    meminfo = mkPluginOpts "meminfo";
+  };
+
+  config = let
+    mkPluginConf = plugin:
+      let pcfg = cfg.${plugin};
+      in optionalString pcfg.enable ''
+        ${plugin} {
+          enabled  = true
+          interval = ${toString pcfg.interval}
+          logmark  = ${toString pcfg.logmark}
+          warning  = ${toString pcfg.warning}
+          critical = ${toString pcfg.critical}
+        }
+    '';
+    watchdogdConf = pkgs.writeText "watchdogd.conf" ''
+      timeout = ${toString cfg.timeout}
+      interval = ${toString cfg.interval}
+      safe-exit = ${boolToString cfg.safeExit}
+      ${mkPluginConf "filenr"}
+      ${mkPluginConf "loadavg"}
+      ${mkPluginConf "meminfo"}
+      ${extraConfig}
+    '';
+  in mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
```

Suggestion: I saw these were mentioned in the descriptions, and I think they are relevant enough to the daemon to be put here.

```suggestion
    systemd.services.watchdogd = {
      documentation = [
        "man:watchdogd(8)"
        "man:watchdogd.conf(5)"
      ];
```

---

Huh! What does that do, where does it show up?
Added.

---

It shows up in (among other things) `systemctl status <service>`. An example would be `systemctl status nix-daemon`, which contains both a manpage and link. It's nice to have when you're staring at a dead systemd unit, as a sort of reminder to go read the docs.